crates/biome_js_type_info/src/local_inference.rs

```diff
@@ -146,72 +152,87 @@ impl Type {
-                        Self::destructuring_of(self.clone(), DestructureField::Name(member_name))
-                            .apply_object_binding_pattern(&pattern),
-                    ),
+                            resolver
+                                .destructuring_of(reference, DestructureField::Name(member_name))
+                                .clone(),
+                        )])
+                    }),
+                    AnyJsBindingPattern::JsArrayBindingPattern(pattern) => Some({
+                        let data = resolver
+                            .destructuring_of(reference, DestructureField::Name(member_name));
+                        data.apply_array_binding_pattern(resolver, &pattern)
+                    }),
+                    AnyJsBindingPattern::JsObjectBindingPattern(pattern) => Some({
+                        let data = resolver
+                            .destructuring_of(reference, DestructureField::Name(member_name));
+                        data.apply_object_binding_pattern(resolver, &pattern)
+                    }),
                 }
             }
-            AnyJsObjectBindingPatternMember::JsObjectBindingPatternShorthandProperty(_) => {
+            AnyJsObjectBindingPatternMember::JsObjectBindingPatternShorthandProperty(_) => Some({
                 let member_name = member_name?;
-                Some(Box::new([(
+                Box::new([(
                     member_name.clone(),
-                    Self::destructuring_of(self.clone(), DestructureField::Name(member_name)),
-                )]))
-            }
-            AnyJsObjectBindingPatternMember::JsObjectBindingPatternRest(rest) => {
+                    resolver.destructuring_of(reference, DestructureField::Name(member_name)),
+                )])
+            }),
+            AnyJsObjectBindingPatternMember::JsObjectBindingPatternRest(rest) => Some({
                 let binding = rest.binding().ok()?;
                 let binding = binding.as_js_identifier_binding()?;
                 let name = text_from_token(binding.name_token())?;
-                Some(Box::new([(
+                Box::new([(
                     name,
-                    Self::destructuring_of(
-                        self.clone(),
+                    resolver.destructuring_of(
+                        reference,
                         DestructureField::RestExcept(names.iter().cloned().collect()),
                     ),
-                )]))
-            }
+                )])
+            }),
             AnyJsObjectBindingPatternMember::JsBogusBinding(_)
             | AnyJsObjectBindingPatternMember::JsMetavariable(_) => None,
         }
     }
 
-    pub fn from_any_js_declaration(decl: &AnyJsDeclaration) -> Self {
+    pub fn array_of(ty: TypeReference) -> Self {
+        Self::instance_of(TypeReference::Qualifier(TypeReferenceQualifier {
+            path: Box::new([Text::Static("Array")]),
+            type_parameters: Box::new([ty]),
```

Comment:
nit: I wonder if we can implement `From` here, to hide the detail that we must pass `Box` in order to make the API happy. Like, `Box` feels like more an implementation detail

Reply:
Looks like `[].into()` also works :+1: Looks a bit cleaner indeed.